### PR TITLE
Fix issue no touch event on mobile

### LINKED ISSUE #37232: Cannot slide images in product page in mobile in 2.4.6 only!
### Preconditions and environment

- Magento version 2.4.6 (in 2.4.5 was working!!)
Luma Theme

### Steps to reproduce

Create a product with more than one image, 3 or 4 would be perfect!
In the product page, using a mobile or mobile view try to slide images to the left (or to the right) with your finger!
Images are not sliding and not changing at all, the slider is not responding. Only way to see the next image is to click on the tiny dot under the image.
I consider this major bug (as usual unexplained regression once again...)

### Expected result

The images should slide following the finger of the user!

### Actual result

Only the first image of the product is displayed and user is completely unable to view the rest images! This is major bug!!!
Why magento is doing so often such basic regressions? Why simple things that used to work correctly are breaking down the road in every release??

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

Delivery patch for recent fix fotorama js issue touch image in product details page.
Feel free to leave any review or comment any your thoughts

Excellent docs from mozilla
https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#improving_scrolling_performance_with_passive_listeners

Logic here
wheel -> passive true cause listener can't block page rendering while the user is scrolling
non wheel (mobile) -> passive false

object options with property passive will pass to event listener at third params

### Fixed Issues (if relevant)

1. Fixes magento/magento2#37232
### Manual testing scenarios (*)

For image type
1. Apply version to 2.4.6
2. Go to any product details (product has media at least 2 images) on mobile device
3. Try to swipe image
4. Image should able to swipe to next image without problems (Previously before patch can't swipe image on mobile)

Also should Re-test for images-videos mixed case

Further tests
Go click on image to enter fullscreen mode
try next image and close fullscreen (X icon) expect works without problems 

PC browser console no errors related with fotorama

Cross-browsers checklists

- Chrome
- Chrome (mobile/touch-events enabled)
- iOS (Chrome & Safari - latest)
- Chrome Android (latest)
- Firefox
 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests B2B, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento give me test instance

**Comment by mrtuvn** [comment]
Another alternative #2 solution

```js
 function addEvent(el, e, fn, bool) {
        if (!e) return;
       let supportsPassive = false;
        try {
          document.addEventListener('test', null, {get passive() {supportsPassive = true}});
        } catch(err) {}
        let options = (supportsPassive) ? {passive: true} : {passive: !!bool};
        el.addEventListener ? el.addEventListener(e, fn, options ) : el.attachEvent('on'+e, fn);
    }

```
Hope this be cleaner for understands

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests look not related to changes from this PR.

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Lima** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Lima** [comment]
@magento give me test instance

**Comment by engcom-Lima** [comment]
@magento give 2.4.6

**Comment by engcom-Lima** [comment]
@magento give me 2.4.6 instance

**Comment by engcom-Lima** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`**  without sample data and **`PHP 8.1`**

**Manual testing scenario:**

- Go to frontend and open product
- Click on image in mobile and swipe right/left

**Before:** :heavy_multiplication_x: Customer was not able to swipe the image.

https://user-images.githubusercontent.com/60198022/227169534-2653cb23-9a56-4a8c-91b3-749cdc848497.mov

https://user-images.githubusercontent.com/60198022/227172486-8388cd40-6cee-45af-a285-1fcb45d9ea45.mov



**After:** :heavy_check_mark:  Customer now able to swipe the image.

https://user-images.githubusercontent.com/60198022/227169704-7d663ff7-f8e3-45a9-bb3a-adaf0adae006.mov

https://user-images.githubusercontent.com/60198022/227172564-46ee5f72-a14c-4901-8afb-a337dde4e0bf.mov

Multiple builds are failed. Hence, moving this PR to **Extended Testing**

**Review by engcom-Lima — CHANGES_REQUESTED** [review]
Hi @mrtuvn ,

Thanks for your contribution and collaboration.

:x: QA not Passed

Video is not getting played in mobile view and getting error in console.

**Manual Scenario Steps:**
- Install fresh magento 2.4-develop without Sample data and apply the fix from this PR.
- Go to frontend in mobile.
- Swipe the images and video
- Click on video.

**:heavy_check_mark: Expected result After Fix:** Video should be played and there should not be any error in the console.

**:x: Actual result After Fix:** Getting error in the console.
Screenshot 2023-03-23 at 4 11 21 PM

**Note:** 
- Video was getting played in mobile view before PR changes.
- I am able to swipe the images and video successfully after the PR changes. 


Thanks

**Comment by mrtuvn** [comment]
Seem another issue from another module. The trace error direct issue in file 
vendor/magento/module-product-video/view/frontend/web/js/fotorama-add-video-events.js
Can you able to reproduce in clean 2.4.6

**Comment by mrtuvn** [comment]
For the issue video at fullscreen mode please check this fix 
https://github.com/magento/magento2/pull/36578/files
cc: @engcom-Lima

File: lib/web/fotorama/fotorama.js [hunk 1]
```diff
@@ -1140,7 +1140,12 @@ fotoramaVersion = '4.6.4';
 
     function addEvent(el, e, fn, bool) {
         if (!e) return;
-        el.addEventListener ? el.addEventListener(e, fn, !!bool) : el.attachEvent('on' + e, fn);
+        const options = {
+            get passive() {
+                return e === 'wheel';
+            }
+        }
+        el.addEventListener ? el.addEventListener(e, fn, options) : el.attachEvent('on' + e, fn);
```
Comment (fredden) [inline lib/web/fotorama/fotorama.js]: Can we drop support for unsupported Internet Explorer v8 yet?
https://caniuse.com/addeventlistener
https://experienceleague.adobe.com/docs/commerce-operations/installation-guide/system-requirements.html#supported-browsers
```suggestion
        el.addEventListener(e, fn, options);
```

Comment (mrtuvn) [reply lib/web/fotorama/fotorama.js]: sure let's try it

Comment (mrtuvn) [reply lib/web/fotorama/fotorama.js]: updated code can you help for reviews

File: lib/web/fotorama/fotorama.js [hunk 2]
```diff
@@ -1140,7 +1140,24 @@ fotoramaVersion = '4.6.4';
...
+
+            document.addEventListener('test', null, options);
+            document.removeEventListener('test', null, options);
+        } catch (err) {}
+
+        const options = (supportsPassive) ? {
+            passive: true
+        } : {
+            passive: !!bool
+        };
```
Comment (fredden) [inline lib/web/fotorama/fotorama.js]: The old code used `!!bool` as the third argument, which is equivalent to `{ capture: !!bool }` not `{ passive: !!bool }`. Is this change intentional?
```suggestion
        } : !!bool;
```

```suggestion
        } : {
            capture: !!bool,
        };
```

File: lib/web/fotorama/fotorama.js [hunk 3]
```diff
@@ -1140,7 +1140,24 @@ fotoramaVersion = '4.6.4';
...
+                    supportsPassive = true;
+                },
+            };
+
+            document.addEventListener('test', null, options);
+            document.removeEventListener('test', null, options);
+        } catch (err) {}
+
+        const options = (supportsPassive) ? {
+            passive: true
```
Comment (fredden) [inline lib/web/fotorama/fotorama.js]: It looks like we lost the `e === 'wheel'` change during a refactor. This is specifically called out in the pull request description, suggesting that it was an important part of the intended changes.

Edit: should we preserve the `capture` setting from the original code?

```suggestion
            passive: e === 'wheel',
            capture: !!bool,
```

File: lib/web/fotorama/fotorama.js [hunk 4]
```diff
@@ -1140,12 +1140,24 @@ fotoramaVersion = '4.6.4';
...
+        try {
+            const options = {
+                get passive() {
+                    supportsPassive = true;
+                },
+            };
+
+            document.addEventListener('test', null, options);
+            document.removeEventListener('test', null, options);
+        } catch (err) {}
```
Comment (fredden) [inline lib/web/fotorama/fotorama.js]: This feature detection is not dependant on any of the arguments to this function (nor any inherited context). Can we move this somewhere else to improve performance / reduce the amount of duplicate work the browser needs to do to detect the feature? I don't know how often / how many times this function (`addEvent()`) is called.

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Unit Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Lima** [comment]
Hi @mrtuvn ,

Thanks for your contribution and collaboration.

I have observed below things before PR changes:
1. Swipe was not happening.
2. Both type of videos (Vimeo + Youtube )were getting played.

**Video for your reference:**
https://shared-assets.adobe.com/link/afaf4496-8d1d-4c77-542c-1899b5609cc4

But when I have taken your PR changes as well as you have suggested [this PR](https://github.com/magento/magento2/pull/36578/files) I have observed below points:
1.  When both Vimeo and Youtube videos also added to the product,then videos were not getting played in the full screen mode.
2. Videos were not playing in the when we click on thumbnails (not in full screen mode)

**Video for your reference:**
https://shared-assets.adobe.com/link/6ca69e7b-cce1-4672-4e8b-e491a82f851b

Please have a look on this and provide us solution of the above observations made.


Thanks

**Comment by ihor-sviziev** [comment]
@engcom-Lima, it looks like a bug, so moving to "needs update"

**Comment by mrtuvn** [comment]
in my opinion problem not related with fotorama core side -> Seem root cause come from component `ProductVideo` as @engcom-Lima mention when mixed product video case. Both cases fails related with video right ?

**Comment by rostilos** [comment]
@mrtuvn If the question is about this PR ( https://github.com/magento/magento2/pull/36578 ): I checked it with your edits, on mobiles in fullscreen mode video will autoplay. Clicking on the thumbnail outside of fullscreen mode does not open the video, on tap nothing happens

**Comment by rostilos** [comment]
It seems to me that the error is still somewhere in the fotorama.js file. I checked with and without your edits. Without your edits stopEvent method is executed with error, click/tap event passes and video is playing. 


With your edits specified method works correctly, but prevents click/tap event, which is weird.

**Comment by mrtuvn** [comment]
I found this line not sure it related with play video ?
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/ProductVideo/view/frontend/web/js/load-player.js#L177

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
https://github.com/magento/magento2/blob/2.4-develop/lib/web/fotorama/fotorama.js#L1522
@glo05363 any reason for set true here ? Is this change intentional? Thanks you

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests, Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@mrtuvn, looks like you were reffering to this commit https://github.com/magento/magento2/commit/d1c5ccfdadbc5667e71770c68e617b2e3ad48b12, and it looks like it was fixing exactly the same issue. Are you sure it's still relevant?

**Comment by mrtuvn** [comment]
@magento give me test instance

File: lib/web/fotorama/fotorama.js [hunk 5]
```diff
@@ -1140,7 +1153,12 @@ fotoramaVersion = '4.6.4';
 
     function addEvent(el, e, fn, bool) {
         if (!e) return;
-        el.addEventListener ? el.addEventListener(e, fn, !!bool) : el.attachEvent('on' + e, fn);
+
+        const options = (Modernizr.passiveeventlisteners) ? {
+            passive: e === 'wheel'
```
Comment (fredden) [inline lib/web/fotorama/fotorama.js]: We seem to have lost the setting for `capture` here.
```suggestion
            passive: e === 'wheel',
            capture: !!bool,
```

**Review by fredden — COMMENTED** [review]
Changes look good. I've left one comment regarding what looks like a regression.

File: lib/web/fotorama/fotorama.js (hunk 5 above)
Comment (mrtuvn) [reply lib/web/fotorama/fotorama.js]: Thanks for your review! Added second param capture

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by cptX** [comment]
Unfortunately I cannot make this patch work in my setup...

**Comment by mrtuvn** [comment]
> The current patch with all these changes is not working. It's impossible to swipe the images. What actually works is taking the original file and implementing only this change:
> 
> ```
> In file lib/web/fotorama/fotorama.js @1143
> 
> -    el.addEventListener ? el.addEventListener(e, fn, !!bool) : el.attachEvent('on' + e, fn);
> +    const options = {
> +        get passive() {
> +            return e === 'wheel';
> +        }
> +    }
> +    el.addEventListener ? el.addEventListener(e, fn, options) : el.attachEvent('on' + e, fn);
> ```

thanks for feedbacks! But latest update also for video mixed case too (if you have multiple images and also video). Have you update with my latest one ? @cptX

**Comment by cptX** [comment]
I don't have videos so I cannot test. But because I have images and images were not sliding I ended up using only this part of the patch...

**Comment by mrtuvn** [comment]
@magento give me test instance

**Comment by mrtuvn** [comment]
After cross check browser i recognised results seem different between chrome - firefox
chrome: can't swipe (not show nav icon overlay like firefox) but you can next to image by dots below thumbnail and can trigger play video
firefox:  show nav icon overlay in thumbnail (click to navigate) -> can able to open video on click thumbnail video

**Comment by cptX** [comment]
I was using firefox and I can ensure that swipe in mobile was not working when using the complete patched fotorama.js file. The only patch that actually worked (and I have it in production now) is the patch @ line 1143 of the original code...

**Comment by mrtuvn** [comment]
@cptX updated my code recently. Give it a try if you need

File: lib/web/fotorama/fotorama.js (hunk 5 above)
Comment (fredden) [reply lib/web/fotorama/fotorama.js]: I can see you've reversed this decision in fa7a9a4bf7733c9c5340680a207811a4d4f3f753. Can you provide some information to explain why?

File: lib/web/fotorama/fotorama.js [hunk 6]
```diff
@@ -1140,7 +1153,13 @@ fotoramaVersion = '4.6.4';
 
     function addEvent(el, e, fn, bool) {
         if (!e) return;
-        el.addEventListener ? el.addEventListener(e, fn, !!bool) : el.attachEvent('on' + e, fn);
+
+        const options = (Modernizr.passiveeventlisteners) ? {
```
Comment (mrtuvn) [inline lib/web/fotorama/fotorama.js:1157]: just curiously do we actual need add passive for every dom element (or just add to body, window object) ? CC: @fredden 
Can we add logic for check specific. Since browsers have updated a lot of changes and magento js system seem fall behind quite far

**Comment by mrtuvn** [comment]
Well seem internal already have update coded so i will mark this PR as draft
Thanks
@cptX you can check code in this here 
https://github.com/magento/magento2/commit/9d751c5a826cab16e55043435406bdb559cc4cdd#diff-fb7ac60e407ee78fc8e5ff56eef13de3f3bbe8a8ce52c14d674f0015f502e7c7

File: lib/web/fotorama/fotorama.js (hunk 6 above)
Comment (mrtuvn) [reply lib/web/fotorama/fotorama.js:1157]: element window.document or window.document.body or window should be set passive false

Comment (fredden) [reply lib/web/fotorama/fotorama.js:1157]: Looking at [this page on `developer.mozilla.org`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#improving_scrolling_performance_with_passive_listeners), it seems that omitting 'passive' seems to be the best option as modern browsers\* default the 'passive' option to a sensible value depending on the event.

\* but not Safari. Safari seems to often lag when implementing features. For example, the 'loading=lazy' attribute for '&lt;img&gt;'.

**Review by fredden — COMMENTED** [review]
I have responded to the query, but GitHub did not dismiss the review request. I'm writing this here in a hope that it will dismiss the review comment.

File: lib/web/fotorama/fotorama.js (hunk 6 above)
Comment (mrtuvn) [reply lib/web/fotorama/fotorama.js:1157]: yeap i think default passive enabled on safari and it's supported well
https://caniuse.com/passive-event-listener
only not support for retired IE browser

**Comment by ihor-sviziev** [comment]
@mrtuvn, since the original issue already closed and marked as fixed, can we close this Pull Request?
